### ISSUE TITLE
Report of INTERNAL ERROR

### ISSUE THREAD

**Issue by evhub** [python/mypy#3574]
Using MyPy 0.511, running `mypy --show-traceback` on
```python
def decorator(func: Callable[..., _T]) -> Callable[..., _T]:
    return func
def func(*args, **kwargs):
    return None
ref = func
@decorator(func)
def func(*args, **kwargs):
    return True
@decorator(func)
def func(*args, **kwargs):
    return func(*args, **kwargs)
```
gives the error
```
.\break_mypy.py:5: error: INTERNAL ERROR -- please report a bug at https://github.com/python/mypy/issues version: 0.511
Traceback (most recent call last):
  File "C:\Anaconda3\Scripts\mypy", line 6, in <module>
    main(__file__)
  File "C:\Anaconda3\lib\site-packages\mypy\main.py", line 46, in main
    res = type_check_only(sources, bin_dir, options)
  File "C:\Anaconda3\lib\site-packages\mypy\main.py", line 93, in type_check_only
    options=options)
  File "C:\Anaconda3\lib\site-packages\mypy\build.py", line 188, in build
    graph = dispatch(sources, manager)
  File "C:\Anaconda3\lib\site-packages\mypy\build.py", line 1595, in dispatch
    process_graph(graph, manager)
  File "C:\Anaconda3\lib\site-packages\mypy\build.py", line 1838, in process_graph
    process_stale_scc(graph, scc, manager)
  File "C:\Anaconda3\lib\site-packages\mypy\build.py", line 1937, in process_stale_scc
    graph[id].type_check_first_pass()
  File "C:\Anaconda3\lib\site-packages\mypy\build.py", line 1510, in type_check_first_pass
    self.type_checker.check_first_pass()
  File "C:\Anaconda3\lib\site-packages\mypy\checker.py", line 177, in check_first_pass
    self.accept(d)
  File "C:\Anaconda3\lib\site-packages\mypy\checker.py", line 264, in accept
    stmt.accept(self)
  File "C:\Anaconda3\lib\site-packages\mypy\nodes.py", line 855, in accept
    return visitor.visit_assignment_stmt(self)
  File "C:\Anaconda3\lib\site-packages\mypy\checker.py", line 1180, in visit_assignment_stmt
    self.check_assignment(s.lvalues[-1], s.rvalue, s.type is None, s.new_syntax)
  File "C:\Anaconda3\lib\site-packages\mypy\checker.py", line 1260, in check_assignment
    self.infer_variable_type(inferred, lvalue, self.expr_checker.accept(rvalue),
  File "C:\Anaconda3\lib\site-packages\mypy\checkexpr.py", line 2058, in accept
    typ = node.accept(self)
  File "C:\Anaconda3\lib\site-packages\mypy\nodes.py", line 1215, in accept
    return visitor.visit_name_expr(self)
  File "C:\Anaconda3\lib\site-packages\mypy\checkexpr.py", line 122, in visit_name_expr
    result = self.analyze_ref_expr(e)
  File "C:\Anaconda3\lib\site-packages\mypy\checkexpr.py", line 160, in analyze_ref_expr
    assert result is not None
AssertionError:
```

**Comment by ethanhs** [comment]
I added this to the provided code, since it wasn't given:
```python
from typing import Callable, TypeVar

_T = TypeVar('_T')
```

This appears to be happening because the last two functions are `OverLoadedFuncDef`, which inherit from FuncBase, and MyPy cannot deduce the type of `func`. If we remove the last function definiton, we get:
```
test.py:9: error: Cannot determine type of 'func'
test.py:10: error: Cannot determine type of 'func'
test.py:13: error: Name 'func' already defined
```
The issue here @evhub is that you aren't supposed to redefine names, to avoid this crash, don't do that. The MyPy bug is that is not caught early enough, and causes the crash.

**Comment by evhub** [comment]
@ethanhs Agreed that the issue is related to improper redefinition of `func`. Nevertheless, MyPy shouldn't report that as `INTERNAL ERROR`.

**Comment by ethanhs** [comment]
Certainly. We need to catch the redefinition earlier or deal with it in another way and not crash. Just wanted to share why it was happening so you could fix it until MyPy is fixed.

I believe we could either modify `FuncBase` to default to `AnyType` for the type, or handle it [in `analyze_ref_expr`](https://github.com/python/mypy/blob/master/mypy/checkexpr.py#L152).

**Comment by gvanrossum** [comment]
Confirmed on master.

**Comment by ethanhs** [comment]
The change here should be pretty easy, but Im not exactly sure which is better, or if a different solution is needed.

**Comment by gvanrossum** [comment]
@ethanhs I'm not sure what change you're referring to -- something in mypy or something in the user's code to work around the crash?

**Comment by ethanhs** [comment]
Sorry, I was referencing my early comment suggesting fixes for Mypy. 
>I believe we could either modify FuncBase to default to AnyType for the type, or handle it in analyze_ref_expr.

By "handle it in analyze_ref_expr" I mean handle the type of the OverloadedFuncDef being None.

**Comment by gvanrossum** [comment]
OK, it'll be up to whoever creates the PR (not me).

**Comment by betaveros** [comment]
For whatever it's worth, I think I also ran into this crash when running mypy on code that boils down to
```python
@overload
def f(a): pass
@overload
def f(a): pass
def f(a): pass
f(0)
```
(my bug being that I failed to import `overload` from `typing`.)

**Comment by ethanhs** [comment]
Yes, it appears that Mypy is incorrectly inferring that any decorator to a function that is defined multiple times means the function is an overload, even if the decorator is not `@overload`! For example, the function item in OverloadedFuncDef has `is_overload==True` for this simple-as-possible code:
```python
@anything
def f(a): pass
def f(a): pass
f(0)
```
I'll take a look at where things go wrong.

**Comment by gvanrossum** [comment]
IIRC the problem is that we need to know whether something is overloaded before we have necessarily resolved whether the decorator refers to `overload` (due to aliasing etc.).

**Comment by ilevkivskyi** [comment]
All the crashes described here are now fixed by #4064 so I am closing this.

(Maybe we need another issue to refactor code around ``OverloadedFuncDef``?)
